Collect/pinot-collect.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH PINOT-COLLECT "1" "March 2007" "pinot-collect - pinot 0.70" "User Commands"
.SH NAME
pinot-collect \- Download an URL from the command-line
.SH SYNOPSIS
.B pinot-collect
[\fIOPTIONS\fR] \fIURL\fR
.SH DESCRIPTION
pinot\-collect \- Download an URL from the command\-line
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH EXAMPLES
.IP
pinot\-collect http://some.website.com/
pinot\-collect xapian:///home/fabrice/.pinot/index/1
.SH "REPORTING BUGS"
Report bugs to fabricecolin@users.berlios.de
.PP
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.


NEWS
2007/03/06 version_0_7_0
Tokenize :
 - now use Dijon's filters (http://dijon.berlios.de/)
Index :
 - support for remote indexes served by xapian-progsrv+ssh
 - index to use by pinot-index specified with --db
Search :
 - date range filtering applies to index searches
 - limit the number of results returned by pinot-search with --max
Daemon :
 - log the daemon's PID
 - sped up unindexing of documents after a directory is deleted
 - caught up with changes in D-Bus 1.0
UI :
 - stored queries can do date range filtering. If dates don't make sense
  (eg From >= To), they are ignored
 - under the Session menu, Status shows various bits of information about
  the indexes and the daemon's crawler
 - prompt for command to use to open documents of a type for which no
  application is defined
 - revamped configuration dialog for external indexes
 - Index > Properties now shows a document's size and number of unique terms
 - most operations that involve peeking at the index are done in the background
 - work-around for desktop files that attempt setting env variables in Exec
 - caught up with changes in D-Bus 1.0
 - Brazilian Portuguese translation by Leonardo Melo
 - Russian translation by Sergey Vostrikov

2006/12/21 version_0_6_5
Build :
 - added option --enable-debug to configure, --enable-soap replaces --with-soap
 - complain bitterly if libtextcat header is not found
 - Pthreads may be provided by a library other than libpthread
 - install configuration files in sysconfdir, libraries in libdir
Tokenize :
 - optimized Ogg/MP3 filter
Index :
 - fixed extraction of language and size from document data
 - store date terms for future date range filtering
 - fixed argument checking in pinot-index
Daemon :
 - new DBus method SetDocumentsLabels to relabel several documents at once
 - process DBus messages in a separate thread, not in the main thread
UI :
 - fixed expansion of .desktop's Exec
 - initialize D-Bus ! Not sure why this didn't cause problems before.
 - fixes for when the locale is not UTF-8
 - on first run, open the Preferences box and show the Indexing tab
 - for stored queries set to index and label results, only apply the new label
  to results that already are in one of the indexes, don't do a full update
 - refresh labelled documents list correctly after properties are changed

2006/12/05 version_0_6_4
Collect :
 - don't needlessly load files that are going to be handled by an helper application
Index :
 - store documents size and file extension, if any
 - fixed concurrency bug that could thrash the index on SMP systems
 - very long capitalized terms could lead to document loss
 - file names were always lower-cased
Search :
 - regenerated the Google SOAP API stubs with gsoap 2.7.8c
UI :
 - added filter on file extension
 - fixed crash on SMP systems when listing an index
 - fixed deadlock when indexing a query's results. Oddly enough, it seems it
  happened only on FreeBSD !
 - Dutch translation by Tikkel
 - Swedish translation by Daniel Nylander
Daemon :
 - set a lower scheduling priority
 - queue events in the database, not in memory
 - clean exit when signalled/stopped while crawling

2006/11/18 version_0_6_3
Collect :
 - watch out for NULL characters in data
Tokenize :
 - fixed memory leak. Temporary documents were not deleted most of the time
UI :
 - prettified results list
 - better abstract highlighting
 - fixed clipboard copy of results list and abstract
Daemon :
 - autostart the daemon process
 - with dbus < 0.70, close the connection

2006/11/04 version_0_6_2
General :
 - query shared-mime-info prefix, so that the applications database can be
  loaded even when Pinot is installed under a different prefix
 - copyright notice was missing in source
Index :
 - detect and support libtextcat 3.0 peculiarities
 - can skip files based on glob pattern
Search :
 - fixed issue where label and directory filters were not applied correctly
  when the filter doesn't start with an upper-case letter. Directory filters
  starting with a non-alphanumeric character only work with Xapian >= 0.9.8.
 - fixed A9, Accoona and Exalead plugins
UI :
 - file patterns to skip can be set in Preferences, Indexing
 - columns showing a timestamp were sorted alphabetically
 - refresh index lists correctly when exiting Preferences
Daemon :
 - fixed major bug where the daemon would loop endlessly reindexing mp3/ogg
  files. When notified that a writable file was closed, check the file was
  actually modified before reindexing it.
 - fixed D-Bus warning about closing the connection when exiting

2006/10/18 version_0_6_1
General :
 - switched to gtk+'s version of xdgmime
Index :
 - can now run queries like "type:text/html and lang:en and (tcp near ip)".
  See README for more information about the syntax and a list of filters.
Search :
 - don't reject MozSearch plugins
 - added plugin for Google code search service
 - filters 'site' and 'file' (host name, file name) apply to Web engines
 - fixed abstract generation
UI :
 - revamped the stored queries editor to allow any number of terms and filters
 - preferences relative to My Documents and My Web Pages gathered under the
  same tab
 - hide the stored queries and engines lists by default to avoid scaring
  people used to Beagle too much ;-)
 - mail accounts configuration wasn't always saved correctly
 - updating a document from the My Web Pages index messed the abstract
Daemon :
 - index attachements in mbox files
 - documents deleted since last crawl were not removed from the index if the
  corresponding location wasn't monitored

2006/09/25 version_0_6_0
Build :
 - modifications for building on Cygwin, contributed by Reini Urban
Monitor :
 - inotify monitor more flexible
SQL :
 - retry if the database is busy
 - save the daemon's crawler history
Index :
 - able to open remote indexes that were not initially available
 - indexes are open in write mode only when necessary
 - prefer Xapian's Flint back-end to Quartz

Collect/pinot-collect.cpp
/*
 *  Copyright 2005,2006 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <iostream>
#include <fstream>

#include "MIMEScanner.h"
#include "Url.h"
#include "DownloaderFactory.h"
#include "config.h"

using namespace std;

static struct option g_longOptions[] = {
	{"help", 0, 0, 'h'},
	{"version", 0, 0, 'v'},
	{0, 0, 0, 0}
};

int main(int argc, char **argv)
{
	int longOptionIndex = 0;

	// Look at the options
	int optionChar = getopt_long(argc, argv, "hv", g_longOptions, &longOptionIndex);
	while (optionChar != -1)
	{
		switch (optionChar)
		{
			case 'h':
				// Help
				cout << "pinot-collect - Download an URL from the command-line\n\n"
					<< "Usage: pinot-collect [OPTIONS] URL\n\n"
					<< "Options:\n"
					<< "  -h, --help		display this help and exit\n"
					<< "  -v, --version		output version information and exit\n"
					<< "\nExamples:\n"
					<< "  pinot-collect http://some.website.com/\n"
					<< "  pinot-collect xapian:///home/fabrice/.pinot/index/1\n"
					<< "\nReport bugs to " << PACKAGE_BUGREPORT << endl;
				return EXIT_SUCCESS;
			case 'v':
				cout << "pinot-collect - " << PACKAGE_STRING << "\n\n"
					<< "This is free software.  You may redistribute copies of it under the terms of\n"
					<< "the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.\n"
					<< "There is NO WARRANTY, to the extent permitted by law." << endl;
				return EXIT_SUCCESS;
			default:
				return EXIT_FAILURE;
		}

		// Next option
		optionChar = getopt_long(argc, argv, "hv", g_longOptions, &longOptionIndex);
	}

	if (argc < 2)
	{
		cerr << "Not enough parameters" << endl;
		return EXIT_FAILURE;
	}

	MIMEScanner::initialize();
	DownloaderInterface::initialize();

	string url(argv[1]);
	Url thisUrl(url);
	cout << "Protocol: " << thisUrl.getProtocol() << endl;
	cout << "User: " << thisUrl.getUser() << endl;
	cout << "Password: " << thisUrl.getPassword() << endl;
	cout << "Host: " << thisUrl.getHost() << endl;
	cout << "Location: " << thisUrl.getLocation() << endl;
	cout << "File: " << thisUrl.getFile() << endl;
	cout << "Parameters: " << thisUrl.getParameters() << endl;

	// Which Downloader ?
	DownloaderInterface *pDownloader = DownloaderFactory::getDownloader(thisUrl.getProtocol());
	if (pDownloader == NULL)
	{
		cerr << "Couldn't obtain downloader for protocol " << thisUrl.getProtocol() << endl;

		DownloaderInterface::shutdown();
		MIMEScanner::shutdown();

		return EXIT_FAILURE;
	}

	DocumentInfo docInfo("Test", url, "", "");
	Document *pDoc = pDownloader->retrieveUrl(docInfo);
	if (pDoc == NULL)
	{
		cerr << "Download operation failed !" << endl;
	}
	else
	{
		cout << "Type: " << pDoc->getType() << endl;

		unsigned int contentLen;
		const char *pContent = pDoc->getData(contentLen);

		if ((pContent != NULL) &&
			(contentLen > 0))
		{
			string fileName(thisUrl.getFile());

			if (fileName.empty() == true)
			{
				fileName = "index.html";
			}

			cout << "Saving " << contentLen << " bytes to " << fileName << endl;

			// Save the content to a file
			ofstream outputFile(fileName.c_str());
			outputFile.write(pContent, contentLen);
			outputFile.close();
		}
		else
		{
			cout << "Document is empty" << endl;
		}

		delete pDoc;
	}

	delete pDownloader;

	DownloaderInterface::shutdown();
	MIMEScanner::shutdown();

	return EXIT_SUCCESS;
}
